src/core/dyad_core.c

```diff
@@ -0,0 +1,622 @@
+        const char **file_data, int *file_len)
+{
+    int rc = -1;
+    flux_future_t *f = NULL;
+    f = flux_rpc_pack(ctx->h,
+            "dyad.fetch",
+            kvs_data->owner_rank,
+            0,
+            "{s:s}",
+            "upath",
+            kvs_data->fpath
+            );
+    if (f == NULL)
+    {
+        DYAD_LOG_ERR(ctx, "Cannot send RPC to producer plugin!\n");
+        return DYAD_BADRPC;
+    }
+    rc = flux_rpc_get_raw(f, (const void**) file_data, file_len);
+    if (f != NULL)
+    {
+        flux_future_destroy(f);
+        f = NULL;
+    }
+    if (rc < 0)
+    {
+        DYAD_LOG_ERR(ctx, "Cannot get file back from plugin via RPC!\n");
+        return DYAD_BADRPC;
+    }
+    return DYAD_OK;
+}
+
+int dyad_pull(dyad_ctx_t *ctx, const char* fname,
+        dyad_kvs_response_t *kvs_data)
+{
+    int rc = validate_context(ctx, true, false);
+    if (DYAD_IS_ERROR(rc))
+    {
+        return rc;
+    }
+    if (ctx == NULL)
+    {
+        fprintf (stderr, "DYAD context not provided to dyad_pull\n");
+        return DYAD_NOCTX;
+    }
+    const char *file_data = NULL;
+    int file_len = 0;
+    const char *odir = NULL;
+    FILE *of = NULL;
+    char file_path [PATH_MAX+1] = {'\0'};
+    char file_path_copy [PATH_MAX+1] = {'\0'};
+
+    rc = dyad_rpc_get(ctx, kvs_data, &file_data, &file_len);
+    if (rc != DYAD_OK)
+    {
+        goto pull_done;
+    }
+
+    ctx->reenter = false;
+
+    strncpy (file_path, ctx->cons_managed_path, PATH_MAX-1);
+    concat_str (file_path, kvs_data->fpath, "/", PATH_MAX);
+    strncpy (file_path_copy, file_path, PATH_MAX); // dirname modifies the arg
+
+    // Create the directory as needed
+    // TODO: Need to be consistent with the mode at the source
+    odir = dirname (file_path_copy);
+    mode_t m = (S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH | S_ISGID);
+    if ((strncmp (odir, ".", strlen(".")) != 0) &&
+            (mkdir_as_needed (odir, m) < 0)) {
+        DYAD_LOG_ERR(ctx, "Cannot create needed directories for pulled file\n");
+        rc = DYAD_BADFIO;
+        goto pull_done;
+    }
+
+    of = fopen(file_path, "w");
+    if (of == NULL)
+    {
+        DYAD_LOG_ERR(ctx, "Cannot open file %s\n", file_path);
+        rc = DYAD_BADFIO;
+        goto pull_done;
+    }
+    size_t written_len = fwrite(file_data, sizeof(char), (size_t) file_len, of);
+    if (written_len != (size_t) file_len)
+    {
+        DYAD_LOG_ERR(ctx, "fwrite of pulled file failed!\n");
+        rc = DYAD_BADFIO;
+        goto pull_done;
+    }
+    rc = fclose(of);
+
+    if (rc != 0)
+    {
+        rc = DYAD_BADFIO;
+        goto pull_done;
+    }
+    rc = DYAD_OK;
+
+pull_done:
+    ctx->reenter = true;
+    if (rc == DYAD_OK && (ctx && ctx->check))
+        setenv (DYAD_CHECK_ENV, "ok", 1);
+    return DYAD_OK;
+}
+
+int dyad_free_kvs_response(dyad_kvs_response_t *resp)
+{
+    if (resp == NULL)
+    {
+        return 0;
+    }
+    free(resp->fpath);
+    free(resp);
+    resp = NULL;
```

Comment:
Do we want to use **resp to make sure resp = NULL to have effect, or is this fine?

Reply:
I'll change to `**resp`